backend/packages/Upgrade/src/api/services/ScheduledJobService.ts

```diff
@@ -6,57 +6,92 @@ import { ScheduledJob, SCHEDULE_TYPE } from '../models/ScheduledJob';
 import { Experiment } from '../models/Experiment';
 import { EXPERIMENT_STATE } from 'upgrade_types';
 import { env } from '../../env';
-import { ExperimentRepository } from '../repositories/ExperimentRepository';
 import { AWSService } from './AWSService';
-import { UserRepository } from '../repositories/UserRepository';
 import { systemUserDoc } from '../../init/seed/systemUser';
 import { ExperimentService } from './ExperimentService';
 import { ErrorRepository } from '../repositories/ErrorRepository';
 import { ExperimentAuditLogRepository } from '../repositories/ExperimentAuditLogRepository';
+import { getConnection } from 'typeorm';
+import { User } from '../models/User';
 
 @Service()
 export class ScheduledJobService {
   constructor(
     @OrmRepository() private scheduledJobRepository: ScheduledJobRepository,
-    @OrmRepository() private experimentRepository: ExperimentRepository,
-    @OrmRepository() private userRepository: UserRepository,
     @OrmRepository() private errorRepository: ErrorRepository,
     @OrmRepository() private experimentAuditLogRepository: ExperimentAuditLogRepository,
     private awsService: AWSService,
     @Logger(__filename) private log: LoggerInterface
   ) {}
 
   public async startExperiment(id: string): Promise<any> {
-    const scheduledJob = await this.scheduledJobRepository.findOne(id, { relations: ['experiment'] });
-    if (scheduledJob && scheduledJob.experiment) {
-      const experiment = await this.experimentRepository.findOne(scheduledJob.experiment.id);
-      if (scheduledJob && experiment) {
-        const systemUser = await this.userRepository.findOne({ email: systemUserDoc.email });
-        const experimentService = Container.get<ExperimentService>(ExperimentService);
-        // update experiment startOn
-        await this.experimentRepository.update({ id: experiment.id }, { startOn: null });
-        return experimentService.updateState(scheduledJob.experiment.id, EXPERIMENT_STATE.ENROLLING, systemUser);
+    return await getConnection().transaction(async (transactionalEntityManager) => {
+      try {
+        const scheduledJob = await transactionalEntityManager.getRepository(ScheduledJob).findOne(id, { relations: ['experiment'] });
+
+        const currentDate = new Date();
+        const timeDif = Math.abs(currentDate.getTime() - scheduledJob.timeStamp.getTime());
+        const fiveHoursInMS = 18000000;
+        
+        if (timeDif > fiveHoursInMS) {
+          const errorMsg =  'Time Differnce of more than 5 hours is found';
+          await transactionalEntityManager.getRepository(ScheduledJob).delete({id: scheduledJob.id});
+          throw new Error(errorMsg);
+        }
+
+        if (scheduledJob && scheduledJob.experiment) {
+          const experiment = await transactionalEntityManager.getRepository(Experiment).findOne(scheduledJob.experiment.id);
```

Comment:
We can extract the Experiment repository only once and then use the same instead of getting every time from the transactionalEntityManager.

Reply:
Code Updated